### PR TITLE
Blocks\Assets\Api: Do not attempt to update script_data transient with the same data

### PR DESCRIPTION
**Pull request by mreishus**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

**On the shutdown hook created by `Blocks\Assets\Api`, only run `set_transient()` if we have new or changed data to put in the transient.** Currently, this is being run on every request, even if the data is exactly the same.

This change aims to improve performance by increasing the TTLB (time to last byte) - which does delay the browser response.

In the overall timeline of a request, this shutdown hook shows near the end:

A closer look:

If we log every SQL query being run, we can see this `set_transient()` call results in 3 queries being run when the data is exactly the same:
```
SELECT autoload FROM wp_options WHERE option_name = '_transient_timeout_woocommerce_blocks_asset_api_script_data' LIMIT 1
SHOW FULL COLUMNS FROM `wp_options`
UPDATE `wp_options` SET `option_value` = '1727274701' WHERE `option_name` = '_transient_timeout_woocommerce_blocks_asset_api_script_data'
```

WP is smart enough to not run the update on the transient option, but this is still 3 queries that don't need to happen.

The timing is variable, but it can take up to 12-15ms in my testing:

Shutdown hooks do impact the browser response. This can be verified by adding a `sleep(10)` to the shutdown hook.

#### Inspect correctness 

Inspect the `$script_data` attribute of the Api.php closely. Besides being pulled out the cache, it is a private attribute, and it is only updated in one place:
https://github.com/woocommerce/woocommerce/blob/30ae60b27336ce050bd69ecf6e2707e7f593a76a/plugins/woocommerce/src/Blocks/Assets/Api.php#L214-L218

Therefore, we only need to set a flag in this one spot to know if `$this->script_data` changed outside the cache and needs to be updated.

#### Verify fewer DB queries

Log out every query being run:
```diff
diff --git a/class-wpdb.php b/class-wpdb.php
index aeb9679..9b20f37 100644
--- a/class-wpdb.php
+++ b/class-wpdb.php
@@ -2255,6 +2255,7 @@ class wpdb {
                // Keep track of the last query for debug.
                $this->last_query = $query;
 
+               error_log($query);
                $this->_do_query( $query );
 
                // Database server has gone away, try to reconnect.
```

Visit the homepage. Notice 3 fewer queries at the very end while using this branch.

#### Verify Script Data

Add: 
```php
		error_log(json_encode([
			'src'          => $this->script_data[ $relative_src ]['src'],
			'version'      => $this->script_data[ $relative_src ]['version'],
			'dependencies' => array_merge( $this->script_data[ $relative_src ]['dependencies'], $dependencies ),
		]));
```
Just before the return statement in `get_script_data()`. Before and after the patch, there should be no changes.

#### Verify cache still working

Add:
```php
		error_log("Found cached script data");
		error_log(json_encode($transient_value['script_data']));
```
Just before the return statement in `get_cached_script_data()`. See that data is being pulled out of the cache.

#### Verify cache updates still happen

Change the return value of `get_script_data_hash()` (can just add a ` . '_v2'` at the end), while logging all DB queries. Your next request should contain a `UPDATE `wp_options` SET `option_value` = '{\"script_data\"...` query. Note that if you used a browser to request, you might have to scroll up in the query log, since your browser made 2 requests and only the first one updated it.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [x] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by alexflorisca — COMMENTED** [review]
Great analysis @mreishus, I've definitely learned a few things looking at this PR! 

My concern with this approach is that the transient should update, not only when the script data is modified, but also the wc version changes. This can happen externally to this file. What do you think would work here?

**Comment by mreishus** [comment]
> Great analysis @mreishus, I've definitely learned a few things looking at this PR!
> 
> My concern with this approach is that the transient should update, not only when the script data is modified, but also the wc version changes. This can happen externally to this file. What do you think would work here?

It will update as long as the `Constants::get_constant( 'WC_VERSION' )` is what's changing. The mechanism is:

* WC_VERSION changing causes `$this->wc_version` to update:

https://github.com/mreishus/woocommerce/blob/30ae60b27336ce050bd69ecf6e2707e7f593a76a/plugins/woocommerce/src/Blocks/Assets/Api.php#L72

* Invalidation mechanism one: get_cached_script_data() returns empty array if the version saved doesn't match the current version:

https://github.com/mreishus/woocommerce/blob/30ae60b27336ce050bd69ecf6e2707e7f593a76a/plugins/woocommerce/src/Blocks/Assets/Api.php#L157

* Invalidation mechanism two: Version is included in the hash and get_cached_script_data() returns empty array if the version saved doesn't match the current version:

https://github.com/mreishus/woocommerce/blob/30ae60b27336ce050bd69ecf6e2707e7f593a76a/plugins/woocommerce/src/Blocks/Assets/Api.php#L136-L138

https://github.com/mreishus/woocommerce/blob/30ae60b27336ce050bd69ecf6e2707e7f593a76a/plugins/woocommerce/src/Blocks/Assets/Api.php#L158-L159

* When get_cached_script_data() returns false, we go through the only modification path available:

https://github.com/mreishus/woocommerce/blob/30ae60b27336ce050bd69ecf6e2707e7f593a76a/plugins/woocommerce/src/Blocks/Assets/Api.php#L205-L208

This is verifiable by adding logging to updates:
```
diff --git a/plugins/woocommerce/src/Blocks/Assets/Api.php b/plugins/woocommerce/src/Blocks/Assets/Api.php
index a501b62170..efa21f3b49 100644
--- a/plugins/woocommerce/src/Blocks/Assets/Api.php
+++ b/plugins/woocommerce/src/Blocks/Assets/Api.php
@@ -181,6 +181,7 @@ class Api {
                if ( ! $this->script_data_modified ) {
                        return;
                }
+               error_log( 'Updating script data cache' );
                set_transient(
                        $this->script_data_transient_key,
                        wp_json_encode(
```

Then changing the version and watching it update once:

```
diff --git a/plugins/woocommerce/src/Blocks/Assets/Api.php b/plugins/woocommerce/src/Blocks/Assets/Api.php
index a501b62170..950e7aa60a 100644
--- a/plugins/woocommerce/src/Blocks/Assets/Api.php
+++ b/plugins/woocommerce/src/Blocks/Assets/Api.php
@@ -76,7 +76,7 @@ class Api {
         */
        public function __construct( Package $package ) {
                // Use wc- prefix here to prevent collisions when WC Core version catches up to a version previously used by the WC Blocks feature plugin.
-               $this->wc_version    = 'wc-' . Constants::get_constant( 'WC_VERSION' );
+               $this->wc_version    = 'wc-' . Constants::get_constant( 'WC_VERSION' ) . 'b';
                $this->package       = $package;
                $this->disable_cache = ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) || ! $this->package->feature()->is_production_environment();

```

**Review by alexflorisca — APPROVED** [review]
Tested this by changing WC version and the `script_data_hash`. Both result in the next request updating the transient, followed by no updates until another change is made. Checked data is still pulled from the cache. Confirm the 3 database queries when no updates are necessary are no longer happening.

For some reason I'm getting 4 calls to the `update_script_data_cache` rather than the expected 2 from 2 browser requests. The last two calls never update the transient, but the first 2 update when necessary. Not sure where this is coming from but I don't think it's a problem, since the updates happen

I've approved this but I would love to get another review here from someone more knowledgeable about the WC internals. Perhaps @opr, @senadir or @mikejolley just for a second sanity check?

Thanks for working on this!

**Review by mikejolley — APPROVED** [review]
I think this makes sense to track 👍🏻 

@alexflorisca if you are merging this, be sure to get a changelog file added beforehand.

**Comment by alexflorisca** [comment]
@kmanijak it would be good if you could take a quick look at this before I merge as well, as it affects all the Woo blocks

File: plugins/woocommerce/changelog/50962-dont-update-transient-with-same-data
Comment (Aljullu) [inline plugins/woocommerce/changelog/50962-dont-update-transient-with-same-data]: Nit: should we mark it as a performance improvement, instead?

```suggestion
Type: performance
```

File: plugins/woocommerce/src/Blocks/Assets/Api.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/Assets/Api.php]: There is a linting error here:

```suggestion
			$this->script_data_modified         = true;
```

**Review by Aljullu — APPROVED** [review]
Thanks for opening this PR, @mreishus! Really nice to see this kind of improvement! I don't have a lot of expertise in this part of the codebase, but code changes look good to me and I did some testing and everything seems to work well, so approving on behalf of Store Editing. :+1:

**Pull request merged**
